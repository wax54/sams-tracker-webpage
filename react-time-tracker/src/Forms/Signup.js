import './Signup.css';

import React, { useContext, useState } from "react";
import { useHistory } from "react-router-dom";
import UserContext from "../UserContext";
import SimpleForm from "./SimpleForm";

const SignupForm  = () => {
    const [ errors, setErrors ] = useState([]);
    const history = useHistory();
    
    const { registerUser } = useContext(UserContext);
    const signupInputs = {
        username : "",
        password : "", 
        firstName : "", 
        lastName : "", 
        email : ""
    }

    const signup = async userData => {

        
        const res = await registerUser(userData);

        if (res.status) {
            history.push('/');
        } else {
            setErrors(res.errors);
        }
    }

    return <SimpleForm 
        className="Signup"
        INITIAL_STATE={signupInputs}
        onSubmit={signup}
        submitText="Sign Up!"
        errors={errors}
    />
};
export default SignupForm