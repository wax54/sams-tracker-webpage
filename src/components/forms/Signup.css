.Signup-error {
    background-color: red;
    color: white;
    margin: 5px;
}
.Signup {
    text-align: center;
}
.Signup * {
    margin: auto;
}