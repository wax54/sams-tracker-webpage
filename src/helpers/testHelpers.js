import { render, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router';
import { Provider } from "react-redux";
import { store } from "../models/store"

function preRender(options = {}) {
    const defaultOptions = {
        initialEntries: ['/'],
    }
    const newOptions = {
        initialEntries:
            [...(options.initialEntries || []), ...defaultOptions.initialEntries],
        store: options.store || store
    }
    //console.log('options: ', options, 'newOptions', newOptions);
    return newOptions;
}

function renderWithRedux(Component, options) {
    options = preRender(options);
    return render(
        <Provider store={options.store}>
            <MemoryRouter initialEntries={options.initialEntries}>
                <Component />
            </MemoryRouter>
        </Provider>
    );
}

async function renderAndWaitFor(Component, options) {
    options = preRender(options);
    return await waitFor(() => render(
        <Provider store={options.store}>
            <MemoryRouter initialEntries={options.initialEntries}>
                <Component />
            </MemoryRouter>
        </Provider>
    ));
}



export {
    renderWithRedux,
    renderAndWaitFor
}