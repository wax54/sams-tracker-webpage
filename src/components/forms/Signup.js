import './Signup.css';


import { useDispatch} from 'react-redux';
import { registerUser } from "../../models/actionCreators";

import React, { useState } from "react";
import SimpleForm from "./SimpleForm";
import { useHistory } from 'react-router-dom';
import { useFormFields } from '../../helpers/hooks';
import { Link } from 'react-router-dom';

const SignupForm = () => {
    const history = useHistory();
    const dispatch = useDispatch();
    const [errors, setErrors] = useState([]);
    const [passwordVisible, toggleVisible] = useState(false);

    
    const signupInputs = {
        username: "",
        password: ""
    }
    const [ {username, password}, handleChange ] = useFormFields(signupInputs);

    const signup = async evt => {
        evt.preventDefault();
        const registered = await dispatch(registerUser({ username, password }));
        if(registered.status) {
            history.push('/');
        } else {
            setErrors([registered.errors]);
        }
    }


    return (
        <form className="Signup container-fluid" onSubmit={signup} >
            {errors.map(message =>
                <div key={message} className="alert alert-danger">
                    {message}
                </div>
            )}
            <div className="input-group input-group-lg my-3">
                <label htmlFor="signup-username"
                    className="input-group-text"
                > Username </label>
                <input
                    type="text"
                    className="form-control"
                    name="username"
                    id="signup-username"
                    value={username}
                    onChange={handleChange}
                />
            </div>
            <div className="input-group input-group-lg my-3">
                <label htmlFor="signup-password"
                    className="input-group-text">
                    Password 
                </label>
                <input
                    type={passwordVisible ? "text" : "password"}
                    className="form-control"

                    name="password"
                    id="signup-password"
                    value={password}
                    onChange={handleChange}
                />
                <label className="input-group-text btn btn-secondary" htmlFor="signup-password" style={{ }} onClick={() => toggleVisible(curr => !curr)} >
                    {passwordVisible ? "Hide" : "View"}
                </label>
            
            </div>
            <input className="btn btn-lg btn-block btn-primary" type="submit" value="Sign Up!"/>
            <Link className="btn btn-block btn-secondary" to="/Login">Already Have An Account?</Link>

        </form>
    )
};
export default SignupForm
