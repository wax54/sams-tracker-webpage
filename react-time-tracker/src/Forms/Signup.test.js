import { fireEvent, waitFor } from '@testing-library/react';
import { renderWithUserContext } from '../testHelpers';
import Signup from './Signup';

//the history object we will be seeing if called
const mockHistory = { push: jest.fn() };
//setting up a mock of the react-router-dom, where almost 
//everything is the same except useHistory returns our mocked history instead
jest.mock('react-router-dom', () => (
    {
        ...jest.requireActual('react-router-dom'),
        useHistory: () => mockHistory,
    }));

describe('Signup Tests', () => {    
    const context = { registerUser: (data) => Promise.resolve({status: true}) }

    test("doesn't blow up", () => {
        renderWithUserContext(Signup, { context });
    });

    test("matchesSnapshot", () => {
        const { asFragment, getByText } = renderWithUserContext(Signup, { context });
        expect(asFragment()).toMatchSnapshot();


        const submitBtn = getByText("Sign Up!");
        fireEvent.click(submitBtn);
    });

    test("gets form data back when submitted", () => {
        const context = {
            registerUser: (data) => {
                result = data;
                return { status: true };
            }
        }

        let result = false;
        const { getByText } = renderWithUserContext(
            Signup, 
            { context });

        const submitBtn = getByText("Sign Up!");
        fireEvent.click(submitBtn);

        expect(result).toEqual({
            username: "",
            password: "",
            firstName: "",
            lastName: "",
            email: ""
        });
    });

    test("redirects to home on success", async () => {

        const { queryByText} = renderWithUserContext(Signup, { context });
        const submitBtn = queryByText("Sign Up!");

        fireEvent.click(submitBtn);
        await waitFor(() => expect(mockHistory.push).toBeCalledWith('/'));
    });

    test("Errors are displayed when they are passed back from registerUser", async () => {
        const { queryByText, debug } = 
            renderWithUserContext(
                Signup, 
                { context: 
                    { registerUser: (data) => Promise.resolve({ status: false, errors: ["DB ERROR"] }) },
                }
            );
        const submitBtn = queryByText("Sign Up!");
        
        expect(queryByText("DB ERROR")).not.toBeInTheDocument();

        fireEvent.click(submitBtn);
        await waitFor(() => expect(queryByText("DB ERROR")).toBeInTheDocument());
    });
    
});
