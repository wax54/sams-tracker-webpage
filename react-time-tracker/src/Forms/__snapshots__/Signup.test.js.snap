// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup Tests matchesSnapshot 1`] = `
<DocumentFragment>
  <form
    class="Signup"
  >
    <div>
      <label
        for="username"
      >
         username: 
      </label>
      <input
        id="username"
        name="username"
        type="text"
        value=""
      />
    </div>
    <div>
      <label
        for="password"
      >
         password: 
      </label>
      <input
        id="password"
        name="password"
        type="text"
        value=""
      />
    </div>
    <div>
      <label
        for="firstName"
      >
         firstName: 
      </label>
      <input
        id="firstName"
        name="firstName"
        type="text"
        value=""
      />
    </div>
    <div>
      <label
        for="lastName"
      >
         lastName: 
      </label>
      <input
        id="lastName"
        name="lastName"
        type="text"
        value=""
      />
    </div>
    <div>
      <label
        for="email"
      >
         email: 
      </label>
      <input
        id="email"
        name="email"
        type="text"
        value=""
      />
    </div>
    <button>
      Sign Up!
    </button>
  </form>
</DocumentFragment>
`;
